workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs

```diff
@@ -33,244 +24,67 @@ public class Worker : IDisposable
 
-        private readonly WorkerDisconnectCallbackSystem disconnectCallbackSystem;
+        // todo replace internal with real apis
+        internal readonly View View;
+        internal ViewDiff ViewDiff;
 
-        /// <summary>
-        ///     An event that triggers when the worker is disconnected.
-        /// </summary>
-        public event Action<string> OnDisconnect
-        {
-            add => disconnectCallbackSystem.OnDisconnected += value;
-            remove => disconnectCallbackSystem.OnDisconnected -= value;
-        }
+        internal MessagesToSend MessagesToSend;
 
-        private Worker(string workerType, Connection connection, ILogDispatcher logDispatcher, Vector3 origin)
+        protected Worker(IConnectionHandler connectionHandler, string workerType, ILogDispatcher logDispatcher)
         {
-            Origin = origin;
+            ConnectionHandler = connectionHandler;
             WorkerType = workerType;
-            WorkerId = connection.GetWorkerId();
+            WorkerId = connectionHandler.GetWorkerId();
+            Attributes = connectionHandler.GetWorkerAttributes();
 
-            Connection = connection;
             LogDispatcher = logDispatcher;
-            logDispatcher.Connection = connection;
+            logDispatcher.Worker = this;
             logDispatcher.WorkerType = workerType;
 
-            World = new World(Connection.GetWorkerId());
-            AddCoreSystems();
-
-            // This isn't a core system, this is for an easy disconnect event
-            disconnectCallbackSystem = World.GetOrCreateManager<WorkerDisconnectCallbackSystem>();
+            MessagesToSend = connectionHandler.GetMessagesToSendContainer();
+            View = new View();
         }
 
-        /// <summary>
-        ///     Tries to connect to the SpatialOS Runtime and creates the worker responsible for the connection upon successfully
-        ///     connecting.
-        /// </summary>
-        /// <param name="connectionFuture">
-        ///     The <see cref="Future{T}" /> of the <see cref="Connection" /> object that we use to
-        ///     connect to the SpatialOS Runtime.
-        /// </param>
-        /// <param name="workerType">The type of the worker.</param>
-        /// <param name="logger">The logger used by this worker.</param>
-        /// <param name="origin">The origin of this worker in the local Unity space.</param>
-        /// <returns>
-        ///     A <see cref="Task{TResult}" /> to run this method asyncally and retrieve the created <see cref="Worker" />
-        ///     object upon connecting successfully.
-        /// </returns>
-        private static async Task<Worker> TryToConnectAsync(Future<Connection> connectionFuture,
-            string workerType,
-            ILogDispatcher logger,
-            Vector3 origin)
+        public static async Task<Worker> CreateWorkerAsync(IConnectionHandlerBuilder connectionHandlerBuilder, string workerType,
```

Comment:
It feels like the connection builder should be extended into a `WorkerBuilder`. I.e. - 

```
var worker = await new WorkerBuilder(workerType)
	.WithConnectionConfig(connectionConfig)
	.WithLogger(logDispatcher)
	.AsMultithreaded()
	.ConnectAsync();
```

Reply:
I didn't do that because you might want different types of connection handler. The IConnectionHandlerBuilder interface is basically a type safe way of transferring ownership.

The current connection handler builder is to build the standard one you might one when running spatial. If you want a test one or a remoting one you want different options. We could write one massive builder that can encompass every useful permutation but that seems like a bad idea even if possible.

Reply:
We could instead set it up so that there is a single connection handler with a pipeline of behaviours and we feed it in different way (ie what it considers the simulation can be rewritten). Then we can potentially just have a single builder and merge it with the worker creation process.

I have looked into doing this and have some rough design notes if you want to go over them. (It's also probably what I'd do if starting now)

Reply:
Pls send design notes

Reply:
Currently writing a doc for a slightly different version. Can go through that and work out how it might work in C# too.